src/ophys_etl/modules/segmentation/graph_utils/feature_vector_rois.py

```diff
@@ -0,0 +1,522 @@
+        global_mask.append(mask)
+    global_mask = np.unique(np.concatenate(global_mask))
+
+    # apply timestep mask
+    sub_video = sub_video[global_mask, :, :]
+    shape = sub_video.shape
+    n_pixels = shape[1]*shape[2]
+    n_time = shape[0]
+    traces = sub_video.reshape(n_time, n_pixels).astype(float)
+    del sub_video
+
+    # mask pixels that need to be masked
+    if pixel_ignore is not None:
+        traces = traces[:, np.logical_not(pixel_ignore.flatten())]
+        n_pixels = np.logical_not(pixel_ignore).sum()
+
+    # calculate the Pearson correlation coefficient between pixels
+    mu = np.mean(traces, axis=0)
+    traces -= mu
+    var = np.mean(traces**2, axis=0)
+    traces = traces.transpose()
+
+    pearson = np.ones((n_pixels,
+                       n_pixels),
+                      dtype=float)
+
+    numerators = np.tensordot(traces, traces, axes=(1, 1))/n_time
+
+    for ii in range(n_pixels):
+        local_numerators = numerators[ii, ii+1:]
+        denominators = np.sqrt(var[ii]*var[ii+1:])
+        p = local_numerators/denominators
+        pearson[ii, ii+1:] = p
+        pearson[ii+1:, ii] = p
+
+    features = np.copy(pearson)
+
+    # subtract off minimum of each feature
+    pearson_mins = np.min(pearson, axis=0)
+    for ii in range(n_pixels):
+        features[:, ii] = features[:, ii]-pearson_mins[ii]
+
+    # Normalize by the interquartile range of each feature.
+    # If the interquartile range is 0, normalize by max-min.
+    # If that is also zero, set the norm to 1.0
+    p75 = np.quantile(features, 0.75, axis=0)
+    p25 = np.quantile(features, 0.25, axis=0)
+    pmax = np.max(features, axis=0)
+    pmin = np.min(features, axis=0)
+
+    feature_norms = p75-p25
+    feature_norms = np.where(feature_norms > 1.0e-20,
+                             feature_norms,
+                             pmax-pmin)
+
+    feature_norms = np.where(feature_norms > 1.0e-20,
+                             feature_norms,
+                             1.0)
+
+    for ii in range(n_pixels):
+        features[:, ii] = features[:, ii]/feature_norms[ii]
+
+    return features
+
+
+class PotentialROI(object):
+    """
+    A class to do the work of finding the pixel mask for a single
+    ROI.
+
+    Parameters
+    ----------
+    seed_pt: Tuple[int, int]
+        The global image coordinates of the point from which to
+        grow the ROI
+
+    origin: Tuple[int, int]
+        The global image coordinates of the upper left pixel in
+        sub_video. (seed_pt[0]-origin[0], seed_pt[1]-origin[1])
+        will give the row, col coordinates of the seed point
+        in sub_video.
+
+    sub_video: np.ndarray
+        The segment of the video data to search for the ROI.
+        Its shape is (n_time, n_rows, n_cols)
+
+    filter_fraction: float
+        The fraction of timesteps to use when correlating pixels
+        (this is used by calculate_pearson_feature_vectors)
+
+    pixel_ignore: Optional[np.ndarray]
```

Comment:
for later, we should consider whether we actually want to do this. for example, I believe HNC was able to find overlapping ROIs that shared pixels just by performing their seed growth clustering emthod from different starting seeds.
maybe we just have a following split/merge step.

Reply:
Sure.

If we do that, we need to be careful to keep track of already masked pixels when detecting peaks in the image. We don't want to get caught in a loop in which the algorithm keeps evaluating the same masked pixels over and over again.